addon/TelemetryRappor.jsm

```diff
@@ -14,205 +14,351 @@ const PREF_RAPPOR_PATH = "toolkit.telemetry.rappor.";
+  for (let i = 0; i < bits; i++) {
+    // Check whether a random number is higher or not than the given probability.
+    let bit = getRandomFloat() < prob;
+    // Calculate the index of the bit to set. This must be done because
+    // we have to set individual bits to one or zero, but what we have are bytes.
+    let idx = Math.floor(i/8);
+    // Set the corresponding bit in the bloom filter to its value. We're using here
+    // the boolean 'bit' as an int (1 if true, 0 if false).
+    arr[idx] |= (bit << (i % 8));
+  }
+  return arr;
+}
+
+/**
+ * Returns a random float between 0 and 1.
+ */
+function getRandomFloat() {
+  // A buffer with just the right size to convert to Float64.
+  let buffer = new ArrayBuffer(8);
+
+  // View it as an Int8Array and fill it with 8 random ints.
+  let ints = new Int8Array(buffer);
+  crypto.getRandomValues(ints);
+
+  // Set the sign (ints[7][7]) to 0 and the exponent (ints[7][6]-[6][5]) to just the
+  // right size (all ones except for the highest bit).
+  ints[7] = 63;
+  ints[6] |= 0xf0;
+
+  // Now view it as a Float64Array, and read the one float from it.
+  return new Float64Array(buffer)[0] - 1;
+}
+
+/**
+ * Create a report.
+ * @param {string} value - Value to encode.
+ * @param {integer} filterSize - Size of the bloom filter in bytes.
+ * @param {integer} numHashFunctions - Number of hash functions.
+ * @param {float} p - Value for probability p.
+ * @param {float} q - Value for probability q.
+ * @param {float} f - Value for probability f.
+ * @param {integer} cohort - Number of cohorts to use.
+ * @param {string} secret - Secret to generate the Permanent Randomized Response.
+ * @param {string} name - Name of the experiment.
+ */
+function createReport(value, filterSize, numHashFunctions, p, q, f, cohort, secret, name) {
+  // Instead of storing a permanent randomized response, we use a PRNG and a stored
+  // secret to re-compute B' on the fly every time we send a report.
+  let bloomFilter = encode(value, filterSize, numHashFunctions, cohort);
+  let prr = getPermanentRandomizedResponse(bloomFilter, f, secret, name);
+  let irr = getInstantaneousRandomizedResponse(prr, p, q);
+  return irr;
 }
 
 var TelemetryRappor = {
+  /**
+   * Receives the parameters for RAPPOR and returns the Instantaneosu Randomized Response.
+   * @param {string} name - Name of the experiment. Used to store the preferences.
+   * @param {string} value v - Value to submit
+   * @param {integer} filterSize k - Size of the bloom filter in bytes.
+   * @param {integer} numHashFunctions h - Number of hash functions.
+   * @param {integer} cohorts m - Number of cohorts to use.
+   * @param {float} f - Value for probability f.
+   * @param {float} p - Value for probability p.
+   * @param {float} q - Value for probability q.
+   *
+   * @return An object containing the cohort and the encoded value in hex.
+   */
+  createReport: function(name, value, filterSize, numHashFunctions, cohorts, f, p, q) {
```

Comment:
To make it clearer what each argument is at the callsite, we can use a different strategy just for this function: let's use an `params` object and pass it as the third parameter. This would become:

`createReport(name, value, params) {`

Within the functions you would access the parameters as e.g. `options.cohorts` and the caller would need to do something like:

`createReport("test", "something.org", {filterSize: 16, numHashFunctions: 2, cohorts: 100, f: 0.0, p:0.35, q: 0.65});`

This would make it clearer which number is what. You would also need to adjust the documentation of this accordingly:

```
/**
 * @param {Object} params - The parameters for the RAPPOR algorithm.
 * @param {string} params.filterSize- Size of the bloom filter in bytes.
```